### ISSUE TITLE
Add Support for Spaces

### ISSUE THREAD

**Issue by zacksiri** [terraform-providers/terraform-provider-digitalocean#42]
DO just got support for spaces. Spaces are essentially S3 buckets but provided by digital ocean.

**Comment by remicaumette** [comment]
Hi, actually DigitalOcean doesn't provide any public api for spaces 😞.
Ref: https://github.com/digitalocean/godo/issues/150

**Comment by hmazter** [comment]
Now it looks like there is an API for spaces.
https://developers.digitalocean.com/documentation/spaces/

**Comment by mikebywater** [comment]
I got the phpleague flysystem working by treating digital ocean spaces exactly as s3, took a while but I think DO have matched the api. I will give it a go tonight

**Comment by lfarnell** [comment]
@radeksimko Due to DigitalOcean providing an S3 compatible API, would it make sense to just mirror the Amazon S3 implementation for what DigitalOcean supports at this time?

**Comment by aknuds1** [comment]
Yes please!

**Comment by abaez** [comment]
If nobody is currently working on this, I would like to take a swing. Really want spaces usage from tf.

**Comment by nreisb** [comment]
@abaez I don't think anyone has taken a crack at this yet. It is definitely on my to-do list and I was planning on sitting down to try and work something out this weekend.

**Comment by abaez** [comment]
@nreisb,

That's no problem with me. I was waiting to hear if it was alright to
work on this. I just want the ability as I'm sure others do as well.

On Wed, Oct 25, 2017, at 14:48, nreisb wrote:
> @abaez[1] I don't think anyone has taken a crack at this yet. It is
> definitely on my to-do list and I was planning on sitting down to try
> and work something out this weekend.> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[2], or mute the thread[3].> 

Links:

  1. https://github.com/abaez
  2. https://github.com/terraform-providers/terraform-provider-digitalocean/issues/42#issuecomment-339432389

**Comment by heartwilltell** [comment]
Hey guys! Any updates about spaces support?

**Comment by slapula** [comment]
I took a stab at adding some partial support in #66. I'm able to spin up and tear down buckets with that resource.  My end goal however is to ultimately get a Spaces backend working for remote state files etc.

**Comment by dmikalova** [comment]
FYI using a DO space as an s3 backend works while we wait for #77 to get merged.

You need to create a **Spaces** [access key](https://cloud.digitalocean.com/account/api/tokens) and a [bucket](https://cloud.digitalocean.com/spaces).

```
terraform {
  backend "s3" {
    skip_requesting_account_id  = true
    skip_credentials_validation = true
    skip_get_ec2_platforms      = true
    skip_metadata_api_check     = true
    access_key                  = ""
    secret_key                  = ""
    endpoint                    = "nyc3.digitaloceanspaces.com"
    region                      = "us-east-1" # Requires any valid AWS region
    bucket                      = "" # Space name
    key                         = "example-folder/terraform.tfstate"
  }
}
```

**Comment by lpil** [comment]
Can this be closed? Or is there something more to do?

**Comment by eddiezane** [comment]
@andrewsomething is this safe to cut a release and close?

We should probably add custom subdomains for Spaces in a separate issue? godo already has support https://github.com/terraform-providers/terraform-provider-digitalocean/issues/203#issuecomment-481645304

**Comment by andrewsomething** [comment]
@eddiezane We'll likely be cutting a release next week. There is some work in-flight.

**Comment by andrewsomething** [comment]
Initial support for Spaces is now available in v1.12.0.  In order to upgrade an existing project to the latest version, run:

```
terraform init -upgrade
```

You can find the full changelog here: https://github.com/terraform-providers/terraform-provider-digitalocean/blob/master/CHANGELOG.md#120-april-23-2019

I'm going to go ahead and close this issue out. Let's track any additional feature requests independently.
